### PR TITLE
feat(c++,spark): support json payload file format

### PR DESCRIPTION
**Pull request by amygbAI**
### Reason for this PR

Fixes for issue #170 added JSON format read & write support for scala and  **read** support for c++ as arrow is yet to support writer

### What changes are included in this PR?

changes to the following files

1. c++ changes in /cpp/include/gar/fwd.h and /cpp/src/filesystem.cc
2. changes in /spark/datasources-32/src/main/scala/org/apache/graphar/datasources/GarDataSource.scala ; /spark/datasources-32/src/main/scala/org/apache/graphar/datasources/GarTable.scala 
3. added folder json and file 
JSONWriterBuilder.scala

### Are these changes tested?

Yes .. some tests are included in the testing@170 folder

### Are there any user-facing changes?

Not sure about this one. Is ability to upload a new file format a user facing one OR a backend one ??

### CODE REVIEW

**Comment by acezen** [comment]
> May we add it [here](https://github.com/apache/incubator-graphar/blob/main/pyspark/graphar_pyspark/enums.py#L57) also?
yes, hi, @amygbAI, could you help to add the JSON file type to pyspark library?

BTW, the CI seems stuck by the code format check, you can fix the format by following command:

c++
```
make graphar-clformat
```

spark
```
mvn spotless:apply
```

**Comment by amygbAI** [comment]
hey folks, have added the **change** required for **pyspark** and also ran the **code formatting** for scala and c++ .. 2 questions
a) is there something similar to be run for pyspark as well ? 
b) do i need to open another pull request ?

**Comment by amygbAI** [comment]
tested .. please carry on

**Review by acezen — CHANGES_REQUESTED** [review]
Thanks for the work,  the change overall LGTM, and there are may have some places to improve:

### Testing
I suggest you generating a ldbc sample graphar data with `json` payload file format as the testing data, you can refer the [example](https://github.com/apache/incubator-graphar/pull/487) to generate the dataset. I think it's easy to just change the https://github.com/apache/incubator-graphar/blob/87d5eeb6a7f86574328da26f08d950e3c5bc3de8/maven-projects/spark/scripts/run-ldbc-sample2graphar.sh#L31
to "json" after rebase to the main.

And then put the generated dataset to grapahr-testing repository.

We put the testing data in the [graphar-testing](https://github.com/apache/incubator-graphar-testing) repository.

you can open a pull request to upload the testing data to the repository and just like other format(csv, parquet and orc)
https://github.com/apache/incubator-graphar-testing/tree/main/ldbc_sample

and then use the dataset to test the feature.

for c++, I suggest just add a test like 
https://github.com/apache/incubator-graphar/blob/87d5eeb6a7f86574328da26f08d950e3c5bc3de8/cpp/test/test_arrow_chunk_reader.cc#L65-L99
but the graph info is load from your upload testing data.

Ditto with Spark:
https://github.com/apache/incubator-graphar/blob/87d5eeb6a7f86574328da26f08d950e3c5bc3de8/maven-projects/spark/graphar/src/test/scala/org/apache/graphar/TestGraphReader.scala#L34-L55

For PySpark, you should add a unit test to test load json data here:
https://github.com/apache/incubator-graphar/blob/main/pyspark/tests/test_reader.py

File: maven-projects/spark/datasources-32/src/main/scala/org/apache/graphar/datasources/json/JSONWriterBuilder.scala [hunk 1]
```diff
@@ -0,0 +1,74 @@
+/* Licensed to the Apache Software Foundation (ASF) under one or more
```
Comment (acezen) [inline maven-projects/spark/datasources-32/src/main/scala/org/apache/graphar/datasources/json/JSONWriterBuilder.scala:1]: The update should apply to both datasources-32 and datasources-33

**Comment by acezen** [comment]
The Spark CI got some error like:
```
Error: ] /home/runner/work/incubator-graphar/incubator-graphar/maven-projects/spark/datasources-32/src/main/scala/org/apache/graphar/datasources/GarTable.scala:42: object JSONWriteBuilder is not a member of package org.apache.graphar.datasources.json
Error: ] /home/runner/work/incubator-graphar/incubator-graphar/maven-projects/spark/datasources-32/src/main/scala/org/apache/graphar/datasources/GarTable.scala:87: class JSONOptions in package json cannot be accessed in package org.apache.spark.sql.catalyst.json
```

Could you double check that the JSONOptions is accessed in package org.apache.spark.sql.catalyst.json?

Note that we run the Spark CI on spark.3.x and spark 3.2.x, the spark 3.4.x code and above may not work.
https://github.com/apache/incubator-graphar/blob/87d5eeb6a7f86574328da26f08d950e3c5bc3de8/.github/workflows/spark.yaml#L47-L52

**Comment by amygbAI** [comment]
thanks for the feedback ! let me get to this by the end of the week ..

**Comment by SemyonSinchenko** [comment]
image
It is private in 3.2, but it is public in the current master.

**Comment by SemyonSinchenko** [comment]
Resolving #493 should resolve an access problem. And it is quite fast task (just rename packages and fix pom). I can do it fast. @acezen what do you think?

**Comment by amygbAI** [comment]
hi @acezen ..how would u like to proceed with JsonOptions case ? should i simply take the 3.4 version from spark directly ?

**Comment by acezen** [comment]
> hi @acezen ..how would u like to proceed with JsonOptions case ? should i simply take the 3.4 version from spark directly ?

How about just borrow the code directly from spark from 3.2 and 3.3 as a temporary solution, since we are now only test spark 3.2 and 3.3. And we can wait to #493 solved, it's ok to remove the borrowed code.

**Comment by amygbAI** [comment]
cool .. let me do that and get back to you folks

**Comment by SemyonSinchenko** [comment]
Let's just wait for #514 and the problem with `private[sql]` will be resolved.

**Comment by acezen** [comment]
> Let's just wait for #514 and the problem with `private[sql]` will be resolved.

#514 has been merged, Hi, @amygbAI , you can rebase the main and check the `private[sql]` problem is resolved or not.

**Comment by amygbAI** [comment]
Hi @acezen, after rebasing, the compilation with changes for datasources-32 ( spark 3.2 ) worked perfectly ! no more trouble with JSONOptions. 

But then i wanted to compile datasources-33 ( spark 3.3.4 ) and it ran into several issues , none related to the changes i made .. i went ahead and checked a few of them and found that the variables its cribbing about dont exist on the spark 3.3 branch BUT are present in 3.4 onwards .. the issue is that if i continue without making changes to datasources-33 the PR will be incomplete ; please advise

```
[ERROR] [Error] /datadrive/APACHE_GRAPHAR/incubator-graphar/maven-projects/spark/datasources-33/src/main/scala/org/apache/spark/sql/graphar/GarScan.scala:141: one more argument than can be applied to method apply: (sqlConf: org.apache.spark.sql.internal.SQLConf, broadcastedConf: org.apache.spark.broadcast.Broadcast[org.apache.spark.util.SerializableConfiguration], dataSchema: org.apache.spark.sql.types.StructType, readDataSchema: org.apache.spark.sql.types.StructType, partitionSchema: org.apache.spark.sql.types.StructType, filters: Array[org.apache.spark.sql.sources.Filter])org.apache.spark.sql.execution.datasources.v2.orc.OrcPartitionReaderFactory in object OrcPartitionReaderFactory
Note that 'aggregation' is not a parameter name of the invoked method.
[ERROR] [Error] /datadrive/APACHE_GRAPHAR/incubator-graphar/maven-projects/spark/datasources-33/src/main/scala/org/apache/spark/sql/graphar/GarScan.scala:182: value LEGACY_PARQUET_NANOS_AS_LONG is not a member of object org.apache.spark.sql.internal.SQLConf
[ERROR] [Error] /datadrive/APACHE_GRAPHAR/incubator-graphar/maven-projects/spark/datasources-33/src/main/scala/org/apache/spark/sql/graphar/GarScan.scala:183: value legacyParquetNanosAsLong is not a member of org.apache.spark.sql.internal.SQLConf
[ERROR] [Error] /datadrive/APACHE_GRAPHAR/incubator-graphar/maven-projects/spark/datasources-33/src/main/scala/org/apache/spark/sql/graphar/GarScan.scala:186: value PARQUET_FIELD_ID_READ_ENABLED is not a member of object org.apache.spark.sql.internal.SQLConf
[ERROR] [Error] /datadrive/APACHE_GRAPHAR/incubator-graphar/maven-projects/spark/datasources-33/src/main/scala/org/apache/spark/sql/graphar/GarScan.scala:187: value parquetFieldIdReadEnabled is not a member of org.apache.spark.sql.internal.SQLConf
[ERROR] [Error] /datadrive/APACHE_GRAPHAR/incubator-graphar/maven-projects/spark/datasources-33/src/main/scala/org/apache/spark/sql/graphar/GarScan.scala:202: one more argument than can be applied to method apply: (sqlConf: org.apache.spark.sql.internal.SQLConf, broadcastedConf: org.apache.spark.broadcast.Broadcast[org.apache.spark.util.SerializableConfiguration], dataSchema: org.apache.spark.sql.types.StructType, readDataSchema: org.apache.spark.sql.types.StructType, partitionSchema: org.apache.spark.sql.types.StructType, filters: Array[org.apache.spark.sql.sources.Filter], parquetOptions: org.apache.spark.sql.execution.datasources.parquet.ParquetOptions)org.apache.spark.sql.execution.datasources.v2.parquet.ParquetPartitionReaderFactory in object ParquetPartitionReaderFactory
Note that 'aggregation' is not a parameter name of the invoked method.
[WARNING] [Warn] /datadrive/APACHE_GRAPHAR/incubator-graphar/maven-projects/spark/datasources-33/src/main/scala/org/apache/spark/sql/graphar/GarScan.scala:244: parameter value sparkSession in method getFilePartitions is never used
[ERROR] [Error] /datadrive/APACHE_GRAPHAR/incubator-graphar/maven-projects/spark/datasources-33/src/main/scala/org/apache/spark/sql/graphar/GarScanBuilder.scala:68: value pushDataFilters is not a member of org.apache.spark.sql.execution.datasources.v2.parquet.ParquetScanBuilder
[ERROR] [Error] /datadrive/APACHE_GRAPHAR/incubator-graphar/maven-projects/spark/datasources-33/src/main/scala/org/apache/spark/sql/graphar/GarScanBuilder.scala:79: value pushDataFilters is not a member of org.apache.spark.sql.execution.datasources.v2.orc.OrcScanBuilder
[ERROR] [Error] /datadrive/APACHE_GRAPHAR/incubator-graphar/maven-projects/spark/datasources-33/src/main/scala/org/apache/spark/sql/graphar/GarScanBuilder.scala:103: not found: value pushedDataFilters
[ERROR] [Error] /datadrive/APACHE_GRAPHAR/incubator-graphar/maven-projects/spark/datasources-33/src/main/scala/org/apache/spark/sql/graphar/GarWriteBuilder.scala:162: not found: value bucketSpec
[WARNING] [Warn] /datadrive/APACHE_GRAPHAR/incubator-graphar/maven-projects/spark/datasources-33/src/main/scala/org/apache/spark/sql/graphar/GarWriteBuilder.scala:107: local val pathName in method validateInputs is never used
[WARNING] [Warn] /datadrive/APACHE_GRAPHAR/incubator-graphar/maven-projects/spark/datasources-33/src/main/scala/org/apache/spark/sql/graphar/GarWriteBuilder.scala:97: parameter value caseSensitiveAnalysis in method validateInputs is never used
[ERROR] [Error] /datadrive/APACHE_GRAPHAR/incubator-graphar/maven-projects/spark/datasources-33/src/main/scala/org/apache/spark/sql/graphar/parquet/ParquetWriteBuilder.scala:114: value PARQUET_FIELD_ID_WRITE_ENABLED is not a member of object org.apache.spark.sql.internal.SQLConf
[ERROR] [Error] /datadrive/APACHE_GRAPHAR/incubator-graphar/maven-projects/spark/datasources-33/src/main/scala/org/apache/spark/sql/graphar/parquet/ParquetWriteBuilder.scala:115: value parquetFieldIdWriteEnabled is not a member of org.apache.spark.sql.internal.SQLConf

```

**Comment by SemyonSinchenko** [comment]
@amygbAI May you push all the latest changes to the branch please?

**Comment by amygbAI** [comment]
done .. rebased using main ( post #514 ) and moved all prior changes ..
compiled using spark 3.2 and 3.3 and tested

**Pull request closed without merging**